runtime/event_loop.cpp

```diff
@@ -38,40 +40,63 @@ bool EventLoop::cancel_async_task(api::Engine *engine, const int32_t id) {
 
 bool EventLoop::has_pending_async_tasks() { return !queue.get().tasks.empty(); }
 
-bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
-  // Loop until no more resolved promises or backend requests are pending.
-  // LOG("Start processing async jobs ...\n");
+void EventLoop::incr_event_loop_interest() { queue.get().interest_cnt++; }
+
+void EventLoop::decr_event_loop_interest() {
+  MOZ_ASSERT(queue.get().interest_cnt > 0);
+  queue.get().interest_cnt--;
+}
+
+inline bool interest_complete() { return queue.get().interest_cnt == 0; }
 
+inline void exit_event_loop() { queue.get().event_loop_running = false; }
+
+bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
+  if (queue.get().event_loop_running) {
+    fprintf(stderr, "cannot run event loop as it is already running");
+    return false;
+  }
```

Comment:
Reentrant event loops can be useful in some scenarios!  For instance, when implementing `alert()`, it's customary to show the dialog, enter the main loop again (so timers and whatnot are processed, but the JS that called the alert function is waiting for the event loop to finish), and leave the event loop when the dialog is dismissed.

Of course, for now, this is perfectly fine, but it's something to keep in mind if such functionality is ever needed.

Reply:
I absolutely agree! And I actually think we'd be able to make the event loop reentrant without too much difficulty, too. I also think that it's ok to punt on that and address it in a future change :)